extensions-contrib/materialized-view-selection/src/main/java/io/druid/query/materializedview/DerivativesManager.java

```diff
@@ -0,0 +1,246 @@
+
+  @LifecycleStart
+  public void start()
+  {
+    log.info("starting derivatives manager.");
+    synchronized (lock) {
+      if (started) {
+        return;
+      }
+      exec = MoreExecutors.listeningDecorator(Execs.scheduledSingleThreaded("SQLMetadataDerivativesManager-Exec--%d"));
+      final Duration delay = config.getPollDuration().toStandardDuration();
+      future = exec.scheduleWithFixedDelay(
+          new Runnable() {
+            @Override
+            public void run() 
+            {
+              try {
+                updateDerivatives();
+              }
+              catch (Exception e) {
+                log.makeAlert(e, "uncaught exception in derivatives manager updating thread").emit();
+              }
+            }
+          },
+          0,
+          delay.getMillis(),
+          TimeUnit.MILLISECONDS
+      );
+      started = true;
+    }
+    log.info("Derivatives manager started.");
+  }
+
+  @LifecycleStop
+  public void stop()
+  {
+    synchronized (lock) {
+      if (!started) {
+        return;
+      }
+      started = false;
+      future.cancel(true);
+      future = null;
+      final ConcurrentHashMap<String, SortedSet<Derivative>> emptyMap = new ConcurrentHashMap<>();
+      ConcurrentHashMap<String, SortedSet<Derivative>> current;
+      do {
+        current = derivativesRef.get();
+      } while (!derivativesRef.compareAndSet(current, emptyMap));
+      exec.shutdownNow();
+      exec = null;
+    }
+  }
+
+  public static ImmutableSortedSet<Derivative> getDerivatives(String datasource)
+  {
+    return ImmutableSortedSet.copyOf(derivativesRef.get().getOrDefault(datasource, Sets.newTreeSet()));
+  }
+
+  public static ImmutableMap getAllDerivatives()
+  {
+    return ImmutableMap.copyOf(derivativesRef.get());
+  }
+
+  private void updateDerivatives()
+  {
+    List<Pair<String, Derivative>> derivativesInDatabase = connector.retryWithHandle(
+        handle ->
+          handle.createQuery(
+              StringUtils.format("SELECT DISTINCT dataSource,commit_metadata_payload FROM %1$s", dbTables.get().getDataSourceTable())
+          )
+              .map(new ResultSetMapper<Pair<String, Derivative>>() 
+              {
+                @Override 
+                public Pair<String, Derivative> map(int index, ResultSet r, StatementContext ctx) throws SQLException 
+                {
+                  String datasourceName = r.getString("dataSource");
+                  try {
+                    DataSourceMetadata payload = objectMapper.readValue(
+                        r.getBytes("commit_metadata_payload"),
+                        DataSourceMetadata.class);
+                    if (!(payload instanceof DerivativeDataSourceMetadata)) {
+                      return null;
+                    }
+                    DerivativeDataSourceMetadata metadata = (DerivativeDataSourceMetadata) payload;
+                    long avgSizePerGranularity = getAvgSizePerGranularity(datasourceName);
```

Comment:
This executes another SQL query while iterating the result of the above SQL. I'm not sure about this because the database connection should be kept until the whole result is consumed. If `getAvgSizePerGranularity()` takes a long time, it may put more pressure on metastore.

Reply:
Probably better to join two tables?

Reply:
In the SQL of `updateDerivatives()`, we get `commit_metadata_payload` and check whether it is a DerivativeDataSourceMetadata. If not, we don't call `getAvgSizePerGranularity()`, so maybe it's not easy to use join to get the target. What do you think if I execute the SQL in `getAvgSizePerGranularity()` after the SQL in `updateDerivatives()` finishes.

Reply:
Sounds good to me. Please go ahead.